Examples cleanup

Mani-Sadhasivam opened the pull request:
This PR is the start of the work focused towards cleaning up the examples in mraa. Following is the list of modifications done:

1. Moved platform specific examples to platform/ subdirectory
2. Modified and moved C examples to c/ subdirectory

Every example has been structured to follow the same coding style.

alext-mkrs commented:
Hi @Mani-Sadhasivam, thanks for your continued contributions first of all :)

I'm not sure the "platform" category makes sense  to me, to be honest - having it together with "c" and "c++" sounds like dividing animals to "four-legged" and "furry", i.e. using orthogonal features. I'd suggest to move them simply to "c", so that the taxonomy is language-based.

I'll also take more specific look at patches later this week.

Mani-Sadhasivam commented:
Hi @alext-mkrs , thanks for your opinion. The reason why I moved platform specific examples to platform/ subdirectory is that, I intended to make the examples more _generic_ so that when a newbie lands on examples/ he can have examples for all peripherals supported by mraa in multiple languages along with some platform specific examples.

If we merge is together with c/ it will make a little bit of confusion. Also, providing a separate subdirectory will help adding other miscellaneous stuffs particular to one platform.

Anyway if you still think on integrating with c/ I am fine with it :)

arfoll commented:
@Mani-Sadhasivam @alext-mkrs - I love the animal dividing reference :) I'm a bit split on this, the problem is if we keep encouraging 'platform' samples the c/ dir will become cluttered, at the same time I like these samples so it's nice to have them. I'm tempted to say @Mani-Sadhasivam has done the work so he can decide, unless @Propanu or someone else has a particularly strong opinion on this.

alext-mkrs commented:
@Mani-Sadhasivam, @arfoll, no problem, I now see the reasoning behind having `platform` dir and basically it makes sense, if only introduces a selection problem for further additions.

Like if I have a C++ example, but platform-specific - do I put it into `c++` or 'platform`? And if the latter - wouldn't that open the door to a messy `platform` dir, where different language-based files will be stored all at once?  Or e.g. my comment regarding `uart_ow` example - I personally don't see anything platform-specific there, but @Mani-Sadhasivam apparently saw it differently and put it there. That sort of thing.

However that's not a big deal for me either, so let's indeed get @Mani-Sadhasivam to decide based on our discussion if no one else has other opinions :)

Mani-Sadhasivam commented on Doxyfile.in:
```diff
@@ -866,6 +866,7 @@ EXCLUDE_SYMBOLS        =
 # command).
 
 EXAMPLE_PATH           = @CMAKE_CURRENT_SOURCE_DIR@/examples/ \
+                         @CMAKE_CURRENT_SOURCE_DIR@/examples/c \
```
Ack. Will squash all 3 commits together

Mani-Sadhasivam commented on examples/i2c_HMC5883L.c:
```diff
@@ -1,136 +0,0 @@
-#include "math.h"
-
-#define MAX_BUFFER_LENGTH 6
-#define HMC5883L_I2C_ADDR 0x1E
```
TBH I don't have the device to test and since I'm doing the cleanup of examples, I wanna make sure all examples are tested and didn't want to nag @arfoll to test this once again after modification.

If this doesn't makes sense to you people, I am happy to drop mine and add this one as `i2c` based example.

alext-mkrs replied:
I'd suggest to have both, actually, the more - the merrier. @arfoll or @Propanu, do you have access to anything HMC5883L-based to do a checkout (I undofrtunately don't)?

arfoll replied:
Haven't had a hmc5883 sensor for a while (well probably somewher :)) - these days I use the grove RGB lcd thing since I know the i2c codes off by heart. Let's leave 2 for now, we can always remove.

Mani-Sadhasivam commented on examples/mmap-io2.c:
```diff
@@ -1,47 +0,0 @@
-    mraa_gpio_context gpio;
-    gpio = mraa_gpio_init(2);
-
-    mraa_gpio_use_mmaped(gpio, 1);
```
Okay. Will add this code block in `#if 0` block and the author as contributor.

arfoll replied:
Uhm this one can be safely removed, I'm planning on dropping the _use_mmaped API with chardev

Mani-Sadhasivam replied:
@arfoll Okay. Thanks for the info!

Mani-Sadhasivam commented on examples/platform/CMakeLists.txt:
```diff
@@ -0,0 +1,17 @@
+include_directories(${PROJECT_SOURCE_DIR}/api/mraa)
+
+if (ONEWIRE)
+  add_executable (uart_ow uart_ow.c)
```
Agree, I mistakenly considered this as a platform specific one. Will move to `c`.

Mani-Sadhasivam commented on examples/initio.c:
```diff
@@ -1,69 +0,0 @@
-/*
- * Author: Houman Brinjcargorabi <houman.brinjcargorabi@intel.com>
```
Sure. Will move this to `platform`

Propanu commented:
LGTM, I don't have any issues with separating platform specific examples into their own folder. My only comment is on the samples.mapping.txt file, since that is used to provide code snippets equivalent to the C/C++ examples for the Java pages. However, that's likely something myself, @arfoll or @alext-mkrs can update once this is ready to be merged.

Mani-Sadhasivam commented:
@alext-mkrs @arfoll @Propanu 

Incorporated all of your valuable review comments. Also, ran `clang-format` on all source files in c/, c++/, platform/ subdirectories. Major change in this changeset is, added `i2c_hmc5883l`, `hellomraa`, `initio` examples back and moved `uart_ow` example to platform/

alext-mkrs commented:
Thanks. I'll take a look on Friday, time permits, or over the weekend.

arfoll commented:
I just pushed this to examples branch with some minor if style fixes - I always wondered why clang-format doesn't find non braced if lines, but it seems i can find them like this ```clang-tidy *.c -fix -checks="readability-braces-around-statements" --``` so I fixed them and threw it in the style commit.

arfoll commented:
Merged - thanks!

delcypher commented:
@arfoll 
> @delcypher - any ideas how/if we can get clang-format to detect these?

I don't think it can and I think that's intentional. `clang-format` is designed to reformat text. All it does is add/remove white space and line breaks (and adds slashes when reformatting macros). It doesn't AFAIK ever add or remove actual code like braces.
